Updated to match currently supported Python/Django/Wagtail versions

jsma opened the pull request:
- Replaces `isort` & `flake8` with `ruff` to follow upstream Wagtail 
- Updates package.json to use latest versions of dependencies 
- Removes Django as a dependency from `pyproject.toml` so that the Django version can be controlled by the installed version of Wagtail
- Adds an option to exclude front-end files from the generated output if not needed

If merged, this would fix #50 and #14

zerolab commented:
@jsma thank you for this. It looks like the tests are failing, could you check that?

zerolab commented on {{ cookiecutter.__project_name_kebab }}/.pre-commit-config.yaml:
```diff
@@ -45,32 +38,32 @@ repos:
           - '@wagtail/stylelint-config-wagtail@0.6.0'
-  - repo: https://github.com/asottile/blacken-docs
-    rev: v1.12.1
+  - repo: https://github.com/adamchainz/blacken-doc
```
```suggestion
  - repo: https://github.com/adamchainz/blacken-docs
```

jsma replied:
Apologies for the copy/paste fails!

jsma commented:
Apologies for not adequately testing this locally. I've fixed the issues and have tested generating a project with and without front-end files and all appears to be working now.

jsma commented:
I had to pre-format some of the Python to avoid any linting issues with `ruff --fix --exit-non-zero-on-fix`. I also updated some of the Github workflow commands to reduce noise in the output (`git init` verbosely asks for an initial branch name, and `pip` wants the latest version installed).

I've tested generating projects with and without front-end, and have manually run the pre-commit bits and everything seems to be passing now. I have _not_ yet added a specific test of the two versions in the github test workflow but can look into that.

FYI, I removed the prompt for the Node version since with the "use_frontend" prompt, cookiecutter doesn't support branching logic to ask the follow up question about their preferred Node version. 

There is a [closed issue](https://github.com/cookiecutter/cookiecutter/issues/913) and an [open PR](https://github.com/cookiecutter/cookiecutter/pull/1645) which ends in a conversation that the cookiecutter project has been forked due to a period of inactivity. 

Setting all of that aside for now, is there any reason to prompt about the Node version? Or can this just default to the latest v18 LTS, which is what upstream Wagtail is using these days?

chris48s commented on {{ cookiecutter.__project_name_kebab }}/.nvmrc:
```diff
@@ -1 +1 @@
-{{ cookiecutter.node_version }}
+18
```
> Setting all of that aside for now, is there any reason to prompt about the Node version? Or can this just default to the latest v18 LTS, which is what upstream Wagtail is using these days?

:+1: Yes. Lets drop node_version + hard-code 18. It is only used in `.nvmrc`

chris48s commented on {{ cookiecutter.__project_name_kebab }}/pyproject.toml:
```diff
@@ -27,30 +27,27 @@ classifiers = [
 ]
 dynamic = ["version"]
 dependencies = [
-    "Django>=3.0,<4.2",
```
I think we should retain at least an explicit minimum version constraint on django, even if we move to an open-ended range. @zerolab - do you have opinions on this one?

zerolab replied:
I agree on the minimum version constraint suggestion

chris48s commented on {{ cookiecutter.__project_name_kebab }}/pyproject.toml:
```diff
@@ -27,30 +27,27 @@ classifiers = [
     "Operating System :: OS Independent",
     "Programming Language :: Python",
     "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3.7",
```
This PR drops python 3.7 from the classifiers in the manifest but 3.7 is still in the tox matrix and the python versions in the CI build. What is the intended minimum compatible python version? 3.7 or 3.8? We should either consistently drop or retain 3.7 compatibility.

It is not a problem introduced in this PR, but this has also made me realise that this `pyproject.toml` does not include a `requires-python` declaration. Lets add an appropriate `requires-python` here while we're working on it.

chris48s replied:
:+1:  Lets make sure we consistently ditch 3.7 then. Here, tox, GH workflows, all the places, and also add a `requires-python = ">=3.8"` to `pyproject.toml`.

chris48s commented on {{ cookiecutter.__project_name_kebab }}/pyproject.toml:
```diff
@@ -27,30 +27,27 @@ classifiers = [
-    "dj-database-url==1.2.0",
-    "pre-commit>=2.21.0,<3.0"
+    "dj-database-url>=2.0.0",
+    "pre-commit"
```
The reason this was pinned to `pre-commit<3.0` is because the cookiecutter generates a package which supports `python>=3.7`. pre-commit 3.0 requires `python>=3.8`. If we're keeping 3.7 compatibility, we need to retain that constraint.

chris48s commented on {{ cookiecutter.__project_name_kebab }}/tox.ini:
```diff
@@ -3,10 +3,8 @@ skipsdist = True
-    python{3.7,3.8,3.9,3.10}-django{4.0}-wagtail{4.0,4.1,main}-{sqlite,postgres}
-    python{3.8,3.9,3.10,3.11}-django{4.1,main}-wagtail{4.1,main}-{sqlite,postgres}
+    python{3.7,3.8,3.9,3.10}-django{3.2}-wagtail{4.1,4.2,5.0,main}-{sqlite,postgres}
+    python{3.8,3.9,3.10,3.11}-django{4.1,4.2,main}-wagtail{4.1,4.2,5.0,main}-{sqlite,postgres}
```
The CI matrixes (postgres and sqlite) in https://github.com/wagtail/cookiecutter-wagtail-package/blob/main/%7B%7B%20cookiecutter.__project_name_kebab%20%7D%7D/.github/workflows/test.yml need to be updated in line with the changes to the tox envs.

chris48s commented on {{ cookiecutter.__project_name_kebab }}/package.json:
```diff
@@ -10,23 +10,23 @@
-    "react": "^16.13.1",
-    "react-dom": "^16.13.1",
-    "styled-components": "^5.1.1"
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
```
There are a number of places where these updates are diverging away from the major versions being used by wagtail core. For example, wagtail core still uses react 16, webpack-cli 4, etc. It feels like we should stay aligned with core on these.

@zerolab can you advise on this one?

zerolab replied:
@lb- any thoughts on this?
My take is we're OK to upgrade some of the tooling in that core will be always more sluggish

jsma replied:
Ah, sorry about that. I got lazy and let PyCharm autocomplete the latest versions, but I should have followed upstream Wagtail instead.

chris48s commented on hooks/post_gen_project.py:
I'm not really fond of writing a bunch of files that we don't want and then deleting them as a post-process. There is another (IMO more correct) way of conditionally including files with cookiecutter, which does not write them in the first place if they are not desired.

For example, if we want to make `package.json` conditional on `cookiecutter.use_frontend == 'y'` we can change the filename for `package.json` to `{% if cookiecutter.use_frontend == 'y' %}package.json{% endif %}`. Cookiecutter will not write files or directories which have an empty name after templating is applied.

All other things being equal, I think that pattern is preferable. Is there anything that deleting files as a post-process is buying us here?

zerolab replied:
🤔
the post-gen hook is a fairly commonly used pattern with cookicutter.
personally I find this cleaner than using filenames like `{% if cookiecutter.use_frontend == 'y' %}package.json{% endif %}` (unless I misread this)

chris48s replied:
Interesting. I think this is the first time I've seen it done this way. Feels conceptually the wrong way round to me, but lets not block this PR on this one. I'm going to resolve it.

jsma replied:
I agree it is clunky. This is my first time working with a cookie-cutter template but the post-process technique is used by `cookiecutter-django` and was the only off the shelf solution I found: https://github.com/cookiecutter/cookiecutter-django/blob/7a35191a76f2ce6b4b0709d004580e17ce626601/hooks/post_gen_project.py#L211-L239

It didn't occur to me that the filename itself could be a template conditional, but given the number of files involved, that might be a different kind of clunky, and it would preclude ever having dependabot inspect package.json.

chris48s commented on {{ cookiecutter.__project_name_kebab }}/setup.cfg:
Why are you deleting all our our linter config?

If ruff is not config-compatible with flake8/isort, we should translate this existing config to ruff's format (or close as possible).
Personally I prefer separating linter config from package manifest (which is why this was in a `setup.cfg`), so my preference would be a `ruff.toml` instead of `pyproject.toml`.

jsma replied:
Ah, I missed adding the `ruff.toml`. I'll take care of that.

chris48s reviewed (changes requested):
Hi. Thanks for having a look at this and sorry it has taken quite a while to get round to reviewing it. One of the reasons I've been putting it off is I needed to carve out a decent chunk of time because there are several distinct changes all happening at once. As a note for future, I'd say

- Bump python/django/wagtail versions
- Conditionally include FE tooling
- Migrate flake8/isort to ruff
- Bump JS packages

should probably have been 4 different PRs, each of which would have been easier to review in isolation. Nevertheless, I have provided a first round of review comments.

jsma commented:
Apologies for dropping the ball on this and thanks to @chris48s for reworking them (including #57)!
